src/ML-POD/compute_podfit.cpp

```diff
@@ -0,0 +1,1564 @@
+{
+  if (narg < 5) error->all(FLERR,"Illegal compute podfit command");
+
+  std::string pod_file = std::string(arg[3]);  // pod input file
+  std::string data_file = std::string(arg[4]); // data input file
+  std::string coeff_file; // coefficient input file
+
+  if (narg > 5)
+    coeff_file = std::string(arg[5]); // coefficient input file
+  else
+    coeff_file = "";
+
+  podptr = new CPOD(lmp, pod_file, coeff_file);
+  read_data_files(data_file, podptr->pod.species);
+
+  if ((int) traindata.data_path.size() > 1)
+    allocate_memory(traindata);
+  else if ((int) testdata.data_path.size() > 1)
+    allocate_memory(testdata);
+
+  // get POD coefficients from an input file
+
+  if (coeff_file != "") podptr->podArrayCopy(desc.c, podptr->pod.coeff, podptr->pod.nd);
+}
+
+CPODFIT::~CPODFIT()
+{
+  traindata.freememory(1);
+  testdata.freememory(1);
+  desc.freememory(1);
+  nb.freememory(1);
+}
+
+void CPODFIT::init()
+{
+
+  if (force->pair == nullptr)
+  error->all(FLERR,"Compute podfit requires a pair style be defined");
+
+  // need an occasional full neighbor list
+
+  neighbor->add_request(this, NeighConst::REQ_FULL | NeighConst::REQ_OCCASIONAL);
+
+  if (modify->get_compute_by_style("podfit").size() > 1 && comm->me == 0)
+    error->warning(FLERR,"More than one compute podfit");
+
+  // compute POD coefficients using least-squares method
+
+  least_squares_fit(traindata);
+
+  // calculate errors for the training data set
+
+  if ((traindata.training_analysis) && ((int) traindata.data_path.size() > 1) )
+    error_analysis(traindata, desc.c);
+
+  // calculate errors for the test data set
+
+  if ((testdata.test_analysis) && ((int) testdata.data_path.size() > 1) && (testdata.data_path != traindata.data_path))
+    error_analysis(testdata, desc.c);
+
+  // calculate energy and force for the training data set
+
+  if ((traindata.training_calculation) && ((int) traindata.data_path.size() > 1) )
+    energyforce_calculation(traindata, desc.c);
+
+  // calculate energy and force for the test data set
+
+  if ((testdata.test_calculation) && ((int) testdata.data_path.size() > 1) && (testdata.data_path != traindata.data_path) )
+    energyforce_calculation(testdata, desc.c);
+
+}
+
+/* ---------------------------------------------------------------------- */
+
+void CPODFIT::init_list(int /*id*/, NeighList *ptr)
+{
+  list = ptr;
+}
+
+template <typename T> void CPODFIT::writearray2file(const char* filename, T *a, int N, int backend)
+{
+
+  if (N > 0){
+
+    FILE *fp = fopen(filename, "wb");
+
+    fwrite( reinterpret_cast<char*>( &a[0] ), sizeof(T) * N, 1, fp);
+
+    fclose(fp);
+  }
+
+  // old version using ofstream:
+  /*
+  if (N>0) {
+    // Open file to read
+    ofstream out(filename, ios::out | ios::binary);
+
+    if (!out) {
+      error->all(FLERR,"Unable to open file ");
+    }
+
+    out.write( reinterpret_cast<char*>( &a[0] ), sizeof(T) * N );
+
+    out.close();
+  }
+  */
+}
+
+bool CPODFIT::is_a_number(std::string line)
```

Comment:
This is redundant. There are "utils::is_integer()" and "utils::is_double()" and for consistent conversion to numbers you can use either the "ValueTokenizer" class or the "utils::numeric()" or "utils:inumeric()" functions.

Reply:
Removed this function.